plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/simulate/SimulateDocumentResult.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.plugin.ingest.transport.simulate;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentBuilderString;
+import org.elasticsearch.ingest.Data;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class SimulateDocumentResult implements Streamable, ToXContent {
+
+    private Data data;
+    private List<SimulateProcessorResult> processorResultList;
+    private Throwable failure;
+
+    public SimulateDocumentResult() {
+
+    }
+
+    public SimulateDocumentResult(Data data) {
+        this.data = data;
+    }
+
+    public SimulateDocumentResult(List<SimulateProcessorResult> processorResultList) {
```

Comment:
I think I would try to have a base abstract class here e.g. `AbstractSimulateDocumentResult`. Then we'd have a sublcass that holds Data and failure, and the other one that holds a list of SimulateProcessorResult. the long readFrom writeTo and toXContent that we have now would be split between the two subclasses. That would make things more readable I think.